src/monitors/turbFlowStats.cpp

```diff
@@ -0,0 +1,268 @@
+                CHKERRQ(ierr);
+
+                if (turbPt) {
+                    // Get the density data from solution point data
+                    PetscReal densLoc;
+                    monitor->densityFunc.function(solPt, &densLoc, monitor->densityFunc.context.get());
+
+                    // Perform actual calculations now
+                    turbPt[monitor->CatOffset.densitySum] += densLoc;
+                    turbPt[monitor->CatOffset.densityDtSum] += densLoc * dt;
+
+                    // March over each field component
+                    for (int p = monitor->fieldTrack[f]; p < monitor->fieldTrack[f + 1]; p++) {
+                        turbPt[monitor->CatOffset.densityMult + p] += fieldPt[p] * densLoc;
+                        turbPt[monitor->CatOffset.densityDtMult + p] += fieldPt[p] * densLoc * dt;
+                        turbPt[monitor->CatOffset.densitySqr + p] += fieldPt[p] * fieldPt[p] * densLoc;
+                        turbPt[monitor->CatOffset.sum + p] += fieldPt[p];
+                        turbPt[monitor->CatOffset.sumSqr + p] += fieldPt[p] * fieldPt[p];
+                        turbPt[monitor->CatOffset.favreAvg + p] += turbPt[monitor->CatOffset.densityDtMult + p] / (turbPt[monitor->CatOffset.densityDtSum] + monitor->tiny);
+                        turbPt[monitor->CatOffset.rms + p] =
+                            PetscSqrtReal(turbPt[monitor->CatOffset.sumSqr + p] / (step + monitor->tiny) - PetscPowReal(turbPt[monitor->CatOffset.sum + p] / (step + monitor->tiny), 2));
+                        turbPt[monitor->CatOffset.mRms + p] = PetscSqrtReal(turbPt[monitor->CatOffset.densitySqr + p] / (turbPt[monitor->CatOffset.densitySum] + monitor->tiny) -
+                                                                            PetscPowReal(turbPt[monitor->CatOffset.densityMult + p] / (turbPt[monitor->CatOffset.densitySum] + monitor->tiny), 2));
+                    }
+                }
+            }
+        }
+        // Free the turbulent flow array
+        ierr = VecRestoreArray(monitor->turbVec, &turbDat);
+        CHKERRQ(ierr);
+    }
+    PetscFunctionReturn(0);
+}
+
+void ablate::monitors::TurbFlowStats::AddField(DM& dm, const char* nameField, PetscInt numComp) {
+    PetscFV fvm;
+    PetscFVCreate(PetscObjectComm(PetscObject(dm)), &fvm) >> checkError;
+    PetscObjectSetName((PetscObject)fvm, nameField) >> checkError;
+    PetscFVSetFromOptions(fvm) >> checkError;
+    PetscFVSetNumComponents(fvm, numComp) >> checkError;
+
+    DMAddField(dm, nullptr, (PetscObject)fvm) >> checkError;
+    PetscFVDestroy(&fvm);
+}
+
+void ablate::monitors::TurbFlowStats::Register(std::shared_ptr<ablate::solver::Solver> solverIn) {
+    // Copy over the solver
+    ablate::monitors::Monitor::Register(solverIn);
+
+    // Copy the master DM over, set fields, and initialize vector values
+    DMPlexFilter(this->GetSolver()->GetSubDomain().GetDM(), this->GetSolver()->GetSubDomain().GetLabel(), 1, &turbDM);
```

Comment:
I think that we may have a memory leak.  In the destructor for this class you may have to destroy the turbid and globVec.